DatadogCore/Sources/Core/Storage/Reading/FileReader.swift

```diff
@@ -30,17 +30,19 @@ internal final class FileReader: Reader {
 
     // MARK: - Reading batches
 
-    func readNextBatch() -> Batch? {
-        guard let file = orchestrator.getReadableFile(excludingFilesNamed: filesRead) else {
-            return nil
+    func readNextBatches(_ limit: Int?) -> [Batch] {
+        let files = orchestrator.getReadableFiles(excludingFilesNamed: filesRead, limit: limit)
+        guard !files.isEmpty else {
+            return []
         }
-
-        do {
-            let dataBlocks = try decode(stream: file.stream())
-            return Batch(dataBlocks: dataBlocks, file: file)
-        } catch {
-            telemetry.error("Failed to read data from file", error: error)
-            return nil
+        return files.compactMap { file in
+            do {
+                let dataBlocks = try decode(stream: file.stream())
+                return Batch(dataBlocks: dataBlocks, file: file)
+            } catch {
+                telemetry.error("Failed to read data from file", error: error)
+                return nil
+            }
```

Comment:
Does it imply reading all files into memory before SDK initiates sending them? Imagine files stored in each of tracks, would that pesimisticaly mean reading up 40 files into memory with `.medium` and up to **400 files** with `.high`?

We've been struggling problems with allocations and steep memory graphs due to batch reading (like https://github.com/DataDog/dd-sdk-ios/pull/1089 or https://github.com/DataDog/dd-sdk-ios/issues/956#issuecomment-1384426485) and this change seems to be a step in opposite direction.

There is no reason to pre-read all files into memory before we start sending first of them. We should read them one-by-one for each upload, keeping the memory graph flat.

Reply:
That's a good callout. I'll modify the implementation to only load single batch at a time